File: src/main/java/com/gitlab/daring/image/command/CommandEnv.java
package com.gitlab.daring.image.command;

import com.gitlab.daring.image.command.structure.Contour;
import com.gitlab.daring.image.features.DMatchResult;
import org.bytedeco.javacpp.opencv_core.KeyPointVector;
import org.bytedeco.javacpp.opencv_core.Mat;
import org.bytedeco.javacpp.opencv_features2d.DescriptorMatcher;
import org.bytedeco.javacpp.opencv_features2d.Feature2D;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CommandEnv {

    public String task = "";
    public String curTask = "";

    public Mat mat = new Mat();
    public Map<String, Mat> mats = new HashMap<>();
    public Map<String, Object> vars = new HashMap<>();
    public List<Contour> contours;

    public Feature2D featureDetector;
    public KeyPointVector keyPoints = new KeyPointVector();
    public DescriptorMatcher descriptorMatcher;
    public DMatchResult matchResult;

    public void setTask(String task) {
        this.task = task; curTask = "";
    }

    public Mat getMat(String key) {
        String k = eval(key);
        return k.isEmpty() ? mat : mats.get(k).clone();
    }

    public void putMat(String key, Mat m) {
        String k = eval(key);
        Mat mc = m.clone();
        if (k.isEmpty()) mat = mc; else mats.put(k, mc);
    }

    public CommandEnv putVar(String key, Object value) {
        vars.put(key, value); return this;
    }

    public String eval(String exp) {
        String vn = exp.startsWith("$") ? exp.substring(1) : "";
        return vn.isEmpty() ? exp : "" + vars.get(vn);
    }

}

File: src/main/kotlin/com/gitlab/daring/image/command/CommandEnv.kt
package com.gitlab.daring.image.command

import com.gitlab.daring.image.command.structure.Contour
import com.gitlab.daring.image.event.ValueEvent
import com.gitlab.daring.image.features.DMatchResult
import org.bytedeco.javacpp.opencv_core.*
import org.bytedeco.javacpp.opencv_features2d.*
import org.bytedeco.javacpp.opencv_xfeatures2d.SURF
import java.util.*

class CommandEnv {

    val startEvent = ValueEvent<CommandEnv>()
    val finishEvent = ValueEvent<CommandEnv>()

    var task = ""
        set(v) { field = v; curTask = "" }

    var curTask = ""

    var mat = Mat()
    var mats = HashMap<String, Mat>()
    var vars = HashMap<String, Any>()
    var contours = emptyList<Contour>()

    var featureDetector: Feature2D = SURF.create()
    var keyPoints = KeyPointVector()
    var descriptorMatcher: DescriptorMatcher = BFMatcher(NORM_L2, true)
    var matchResult: DMatchResult? = null

    fun getMat(key: String): Mat {
        val k = eval(key)
        return if (k.isEmpty()) mat else mats[k]!!.clone()
    }

    fun putMat(key: String, m: Mat): CommandEnv {
        val k = eval(key)
        val mc = m.clone()
        if (k.isEmpty()) mat = mc else mats.put(k, mc)
        return this
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> getVar(key: String) = vars[key] as T

    fun putVar(key: String, value: Any): CommandEnv {
        return apply { vars.put(key, value) }
    }

    fun eval(exp: String): String {
        val vn = if (exp.startsWith("$")) exp.substring(1) else ""
        return if (vn.isEmpty()) exp else "" + vars[vn]
    }

}
